apps/frontend/src/lib/ws/index.ts

```diff
@@ -0,0 +1,141 @@
+
+  if (expectConnectResp && type !== WebSocketRequestTypes.CONNECT) {
+    console.warn("While waiting for successful user connection response no other websocket requests are allowed, abort.");
+    return;
+  } else if (!successfulConnection && type !== WebSocketRequestTypes.CONNECT) {
+    console.error("Websocket user connection was not successfully established, abort.");
+    return;
+  }
+
+  const req: WebSocketRequest = {
+    authToken,
+    type,
+    targetSigPubKey,
+    senderSigPubKey,
+    payload,
+  }
+
+  wsClient.send(JSON.stringify(req));
+  return;
+}
+
+export const wsConnectRequest = (authToken: string, senderSigPubKey: string): void => {
+  wsRequest(authToken, WebSocketRequestTypes.CONNECT, "", senderSigPubKey, "");
+  return;
+}
+
+// Call in a context where localstorage is available client-side.
+// localstorage cannot be used with server-side rendering.
+export const WebsocketConnectUser = async (): Promise<void> => {
+  if (successfulConnection) {
+    return;
+  }
+
+  try {
+    const user = await storage.getUser();
+    const session = await storage.getSession();
+
+    if (user) {
+      const sender: string = getUserSigPubKey(user);
+      if (sender && session) {
+        // First step after creating connection is sending server sender public signing key for client socket lookup
+        wsConnectRequest(session.authTokenValue, sender);
+
+        // Expect CONNECT_SUCCESS response
+        expectConnectResp = true;
+      }
+    } else {
+      throw new Error("No user public signing key available, unable to establish websocket connection.");
+    }
+  } catch (error) {
+    console.error(`Onopen websocket client error: ${errorToString(error)}`);
+    return;
+  }
+}
+
+export const WebsocketCloseUser = async (): Promise<void> => {
+  try {
+    const { user, session } = await storage.getUserAndSession();
+    const senderSigPubKey: string = getUserSigPubKey(user);
+
+    // Close server's client socket
+    wsRequest(session.authTokenValue, WebSocketRequestTypes.CLOSE, "",  senderSigPubKey, "");
+  } catch (error) {
+    console.error("Onclose socket client error:", errorToString(error));
+    return;
+  }
+}
+
+export const wsClient: WebSocket = new WebSocket(`${BASE_API_WS}`);
+
+wsClient.onopen = async () => {
+  console.log("Open websocket connection");
+};
+
+wsClient.onmessage = async (ev: IMessageEvent) => {
+  if (!ev || !ev.data) {
+    console.warn("Websocket message is invalid");
+    return;
+  }
+
+  try {
+    const resp: WebSocketResponse = WebSocketResponseSchema.parse(JSON.parse(ev.data.toString()));
+
+    switch (resp.type) {
+      case WebSocketResponseTypes.SUCCESS_CONNECT:
+        // Enable other message types now
+        expectConnectResp = false;
+        successfulConnection = true;
+        return;
+      case WebSocketResponseTypes.MSG:
+        const createMessage: CreateMessageData = CreateMessageDataSchema.parse(resp.payload);
```

Comment:
websocket messages should not be tied to the Message schema, they should be separate

Reply:
Yup, this will be removed, it was involved because I thought WS was being used for low latency transfer, as opposed to notifying the client to pull.